File: 120. Triangle/120. Triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        # Get last row as initial dp array
        dp = triangle[-1][:]
        
        # Start from second to last row and move up
        for i in range(len(triangle) - 2, -1, -1):
            for j in range(len(triangle[i])):
                # For each position, take current value + min of two possible paths below
                dp[j] = triangle[i][j] + min(dp[j], dp[j + 1])
        
        return dp[0]

File: 120. Triangle/120. Triangle.dart
class Solution {
  int minimumTotal(List<List<int>> triangle) {
    List<int> dp = List.from(triangle.last);
    
    for (int i = triangle.length - 2; i >= 0; i--) {
      for (int j = 0; j <= i; j++) {
        dp[j] = triangle[i][j] + min(dp[j], dp[j + 1]);
      }
    }
    
    return dp[0];
  }
}